src/NUnitFramework/tests/Constraints/ConstraintCastingTests.cs

```diff
@@ -0,0 +1,203 @@
+using System.Collections.Generic;
+using System.Linq;
+
+#pragma warning disable NUnit2046 // Use Constraint for better assertion messages in case of failure
+namespace NUnit.Framework.Tests.Constraints
+{
+    /// <summary>
+    /// A test fixture that ensures constraints cast their associated property,
+    /// collection item, etc. to its actual type instead of just <see cref="object"/>.
+    /// </summary>
+    [TestFixture]
+    public class ConstraintCastingTests
```

Comment:
The issue #4998 has some example code. 
Could this code be added as a Test here ?

```csharp

   var actual = new Foo
    {
        Bar = 0
    };

    Assert.Multiple(() =>
    {
        // passes
        Assert.That(actual.Bar, Is.Default);

        // fails
        Assert.That(actual, Has.Property(nameof(Foo.Bar)).Default);
    });
    
```

Reply:
The tests added `LengthPropertyConstraint` and `CountPropertyConstraint` cover the same functionality:
```csharp
Assert.That(defaultLength.Length, Is.Default);
Assert.That(defaultLength, Has.Length.Default);
Assert.That(defaultLength, Has.Property("Length").Default);
```

I don't see the need to add a `Foo` class just for the sake of it. I rather update #4998 to use `Length` as an example.